### DISCUSSION TITLE
Retro-fitting AuthP in an existing multi-tenant database

### DISCUSSION THREAD

**Discussion by diddlys** [JonPSmith/AuthPermissions.AspNetCore#88] [Q&A]
I have only recently come across the AuthP library, having got here via various routes including OrchardCore and ABP. Those frameworks have features I like, though neither seem to cater for hierachical multi-tenancy as far as I can see.
We have an existing hierarchical multi-tenant app, which uses combinations of SQL query structures (to query for tenant data 'ownership'), roles and claims to deliver a multi-tenant hierarchy of sorts. We do not currently have the concept of a DataKey in the database, and I have a few questions:
1. Is a DataKey needed in all tables where tenant-owned (and therefore data private to a specific tenant) data is stored? In our case, this is many tens of tables.
2.  Is retro-fitting AuthP feasible into a database that does not have any DataKeys currently?
Thanks for any help. I should say that I have not yet had a play with the sample code, so this may all be apparent when I do that but if anyone could help me out I would be grateful.

**Comment by JonPSmith** [comment]
Hi @diddlys,

I assume you are using schema-per-tenant approach, which is why you don't have a DataKey (although I'm not sure how you can do hierarchical multi-tenants without a DataKey  - see my comments later). Unfortunately, the AuthP library, which uses EF Core (which I know very well due to writing the book [Entity Framework Core in Action](https://bit.ly/EfCoreBookEd2)) needs a DataKey to differentiate each "tenant" (_NOTE: if using [sharding](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Multi-tenant-explained#2-defining-the-three-multi-tenant-database-arrangements), then you don't need a DataKey_). That means you would need to do change your database to add a DataKey (and have one schema), which would be a a big task. 

Also, unlike OrchardCore, the AuthP library is backend code, so you have to build the frontend code. To be honest, if I was doing this I would build a new app using AuthP and when its working, then I would build some code to read the OrchardCore database and setup the tenants, users etc. - that's a lot of work.

> though neither seem to cater for hierachical multi-tenancy as far as I can see.

I'm not sure what you mean by "hierarchical multi-tenancy", but AuthP has quite good hierarchical multi-tenancy feature - see the article ["Building ASP.NET Core and EF Core hierarchical multi-tenant apps"](https://www.thereformedprogrammer.net/building-asp-net-core-and-ef-core-hierarchical-multi-tenant-apps/). The [Example4](https://github.com/JonPSmith/AuthPermissions.AspNetCore/tree/main/Example4.MvcWebApp.IndividualAccounts) app in the repo provides a example hierarchical multi-tenant - each example is designed to load test data, users etc. on first startup so that you can try out its features.
